xpublish/rest.py

```diff
@@ -137,17 +140,27 @@ def extract_zattrs(da):
         zattrs[k] = _encode_zarr_attr_value(v)
     zattrs[_DIMENSION_KEY] = list(da.dims)
 
+    # We don't want `_FillValue` in `.zattrs`
+    # It should go in `fill_value` section of `.zarray`
+    _ = zattrs.pop('_FillValue', None)
+
     return zattrs
 
 
+def _extract_fill_value(da, dtype):
+    da = encode_zarr_variable(da)
```

Comment:
we end up calling this function twice on each data array (here and in line 49). Can we restructure this to use `encoded_da`?

Reply:
We can definitely make some changes. The trick part is getting the original dtype. In those cases where `mask_and_scale` has been applied to the array, it's my understanding that we want to use the `dtype` from `da.encoding` instead of `da.dtype`. Moreover, `xarray.backends.zarr.encode_zarr_variable()` throws away some crucial information. Notice how the `dtype` key is not longer available in the encoding of the array:

```python
In [80]: ds.Tair.encoding                                                                                          
Out[80]: 
{'source': '/Users/abanihi/.xarray_tutorial_data/rasm.nc',
 'original_shape': (36, 205, 275),
 'dtype': dtype('>f8'),
 '_FillValue': 9.969209968386869e+36,
 'coordinates': 'yc xc'}

In [81]: xr.backends.zarr.encode_zarr_variable(ds.Tair).encoding                                                   
Out[81]: 
{'source': '/Users/abanihi/.xarray_tutorial_data/rasm.nc',
 'original_shape': (36, 205, 275),
 'coordinates': 'yc xc'}

In [82]: xr.backends.zarr.encode_zarr_variable(ds.Tair).attrs                                                      
Out[82]: 
{'units': 'C',
 'long_name': 'Surface air temperature',
 'type_preferred': 'double',
 'time_rep': 'instantaneous',
 '_FillValue': 9.969209968386869e+36}
```


One solution is to add `dtype` argument to `extract_zarray()`. Would the following be a reasonable change?:

```python
def extract_zarray(da, encoding, dtype):
    # TODO: do a better job of validating some of these
    meta = {
        'compressor': encoding.get('compressor', da.encoding.get('compressor', default_compressor)),
        'filters': encoding.get('filters', da.encoding.get('filters', None)),
        'chunks': encoding.get('chunks', None),
        'dtype': dtype.str,
        'fill_value': _extract_fill_value(da, dtype),
        'order': 'C',
        'shape': list(normalize_shape(da.shape)),
        'zarr_format': zarr_format,
    }
....
```
Which gets called via 
```python
zmeta['metadata'][f'{key}/{array_meta_key}'] = extract_zarray(
                encoded_da, self._encoding.get(key, {}), da.encoding['dtype']
            )
```

Reply:
We can come back to this if it turns out to be a big deal.

Reply:
Sounds good